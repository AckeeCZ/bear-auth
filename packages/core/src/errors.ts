const errorCodes = [
    'bear-auth/unique-instance',
    'bear-auth/not-implemented',
    'bear-auth/unknown-instance',
    'bear-auth/unset-hook',
    'bear-auth/not-authenticated',
    'bear-auth/refresh-token-failed',
    'bear-auth/fetch-auth-data-failed',
    'bear-auth/logout-failed',
    'bear-auth/unknown-hook',
    'bear-auth/retrieve-auth-session-failed',
    'bear-auth/invalid-expires-in',
] as const;

type ErrorCodes = typeof errorCodes;

type ErrorCode = ErrorCodes[number];

export class BearAuthError<Code extends ErrorCode> extends Error {
    code: ErrorCode;
    originalError?: unknown;

    constructor(code: Code, message: string, error?: unknown) {
        super(message);
        this.name = 'BearAuthError';
        this.code = code;
        this.originalError = error;
    }
}

/**
 * A TS guard function to recognize BearAuthError.
 */
export const isBearAuthError = (error: unknown): error is BearAuthError<ErrorCode> =>
    error instanceof BearAuthError && error.name === 'BearAuthError';

const errorCodeValues = new Set(errorCodes);

/**
 * Specify BearAuthError with specific error codes you want to handle.
 */
export const isBearAuthErrorWithCodes = <const Codes extends ErrorCode[]>(
    error: unknown,
    codes: Codes,
): error is BearAuthError<Codes[number]> => {
    return isBearAuthError(error) && codes.every(code => errorCodeValues.has(code));
};
