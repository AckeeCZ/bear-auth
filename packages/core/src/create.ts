import { createDefaultAlarmManager, type AlarmManager } from './alarmManager.ts';
import {
    setDefaultAuthSessionPropagation,
    type AuthSessionPropagationType,
} from './authSessionPropagation/authSessionPropagation.ts';
import { BearAuthError } from './errors.ts';
import { instances } from './instances.ts';
import { createDefaultLogger, defaultLogLevel, type Logger, type LogLevel } from './logger.ts';
import { defaultContinueWhenOnline } from './network.ts';
import type { OnAuthStateChangedCallback } from './onAuthStateChanged.ts';
import type { StorageSchema } from './storage.ts';
import type { AuthenticatedSession, RefreshingSession, Session } from './store/session.ts';
import { createStore } from './store/store.ts';

export type BearAuth<AuthInfo> = {
    id: string;

    /**
     * It's null if the custom logger is set.
     * @default 'error'
     */
    loglevel: LogLevel | null;
    logger: Logger;

    hooks: {
        refreshToken: null | ((authSession?: RefreshingSession<AuthInfo>['data']) => Promise<Session<AuthInfo>>);
        fetchAuthInfo: null | ((authSession?: AuthenticatedSession<AuthInfo>['data']) => Promise<Session<AuthInfo>>);
        logout: null | (() => Promise<void>);
    };
    store: ReturnType<typeof createStore<AuthInfo>>;
    flags: {
        autoRefreshAccessTokenEnabled: boolean;
        customLogger: boolean;
    };

    storage: StorageSchema<AuthInfo> | null;

    /**
     * Internal storage version. If increased, then during `start` method, the storage will be cleared.
     */
    storageVersion: 1;

    onAuthStateChanged: {
        prevSession: Session<AuthInfo> | null;
        callbacks: Set<OnAuthStateChangedCallback<AuthInfo>>;
    };

    refreshTokenTimeoutId: null | number | string;

    /**
     * @param taskName - The name of the task that is being executed.
     * @returns A promise that resolves when the network is online.
     */
    continueWhenOnline: (
        taskName: 'retrieveAuthSession' | 'logout' | 'refreshToken' | 'fetchAuthInfo',
    ) => Promise<void>;

    authSessionPropagation: {
        type: Exclude<AuthSessionPropagationType, 'none'>;
        cleanUp: () => void;
    } | null;

    alarmManager: AlarmManager;
};

export interface CreateProps {
    /**
     * Choose a unique instance ID when creating a multiple BearAuth instances.
     */
    id?: string;
}

/**
 * Initialize BearAuth instance. Can't be called multiple times with the same instance ID (Call `destroy` first).
 * @returns a string reference to the created BearAuth instance.
 */
export function create({ id = 'bear_auth' }: CreateProps = {}) {
    if (instances.has(id)) {
        throw new BearAuthError(
            'bear-auth/unique-instance',
            `BearAuth with '${id}' ID has been already created. Call 'await detroy(bearAuth: BearAuth)' first.`,
        );
    }

    const instance = {
        id,

        loglevel: defaultLogLevel,
        logger: createDefaultLogger(id, defaultLogLevel),

        hooks: {
            refreshToken: null,
            fetchAuthInfo: null,
            logout: null,
        },

        store: createStore<unknown>(),

        flags: {
            autoRefreshAccessTokenEnabled: false,
            customLogger: false,
        },

        storage: null,

        storageVersion: 1,

        onAuthStateChanged: {
            prevSession: null,
            callbacks: new Set<OnAuthStateChangedCallback<unknown>>(),
        },

        refreshTokenTimeoutId: null,

        continueWhenOnline: defaultContinueWhenOnline,

        authSessionPropagation: null,

        alarmManager: createDefaultAlarmManager(),
    } as const satisfies BearAuth<unknown>;

    instances.set(id, instance);

    setDefaultAuthSessionPropagation(id);

    return instance.id;
}
