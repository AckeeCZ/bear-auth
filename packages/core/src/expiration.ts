import { BearAuthError } from './errors';

/**
 * Convert the `expiresIn` value in ms to a ISO timestamp representing the expiration time.
 */
export function getExpirationTimestamp(expiresIn: number): string {
    return new Date(Date.now() + expiresIn).toISOString();
}

/**
 * Buffer to account for network latency
 */
const REFRESH_TOKEN_EXPIRATION_BUFFER = 5_000; // 5s

const MIN_EXPIRES_IN = 10_000; // 10s

/**
 * Expiration timestamp must be at least 10s (refering to the `MIN_EXPIRES_IN` constant) in the future.
 */
function isValidExpirationTimestamp(expiration: string | undefined | null): expiration is string {
    if (!expiration) {
        return false;
    }

    return Date.now() + MIN_EXPIRES_IN <= Date.parse(expiration);
}

/**
 * Decrease the expiration timestamp by `REFRESH_TOKEN_EXPIRATION_BUFFER` ms (i.e. to account for network latency during access token refresh).
 * @param expiration expiration timestamp in ISO format
 * @returns decreased expiration timestamp in ISO format or `null` if the expiration timestamp too low or is invalid
 */
function decreaseExpirationTimestamp(expiration: string): string | null {
    const decreasedExpiration = new Date(Date.parse(expiration) - REFRESH_TOKEN_EXPIRATION_BUFFER).toISOString();

    return isValidExpirationTimestamp(decreasedExpiration) ? decreasedExpiration : null;
}

/**
 * If expiration timestamp provided, it must be at least 10s in the future (referring to the `MIN_EXPIRES_IN` constant) or the BearAuthError<'bear-auth/invalid-expires-in'> is thrown.
 * If it's provided and valid, the expiration timestamp is decreased by `REFRESH_TOKEN_EXPIRATION_BUFFER` ms.
 * If not provided, `null` is returned.
 * @param expiration
 * @returns
 */
export function getExpirationTimestampWithBuffer(expiration: string | undefined | null) {
    const isValid = isValidExpirationTimestamp(expiration);

    if (expiration && !isValid) {
        throw new BearAuthError(
            'bear-auth/invalid-expires-in',
            `Expiration timestamp must be greater than '${new Date(Date.now() + MIN_EXPIRES_IN).toISOString()}', received: '${expiration}'.}`,
        );
    }

    return isValidExpirationTimestamp(expiration) ? decreaseExpirationTimestamp(expiration) : null;
}
