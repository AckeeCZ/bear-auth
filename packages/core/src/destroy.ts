import { stopTokenAutoRefresh } from './autoRefreshToken.ts';
import { type BearAuth } from './create.ts';
import { getInstance, instances } from './instances.ts';
import { runOnAuthStateChangedCallbacks } from './onAuthStateChanged.ts';
import { setUnauthenticatedSession } from './store/session.ts';
import { allTasksSettled, deleteInstanceTasks } from './tasks.ts';

/**
 * Destroys the BearAuth instance.
 * - Tries to logout user.
 * - Clears the storage.
 * - Deletes the instance.
 * @param id - return value of `create` method
 */
export async function destroy<AuthInfo>(id: BearAuth<AuthInfo>['id']) {
    const instance = getInstance<AuthInfo>(id);

    if (!instance) {
        return;
    }

    await stopTokenAutoRefresh<AuthInfo>(id);

    const { logger, authSessionPropagation, hooks, storage, store } = instance;

    logger.debug('[destroy]', 'Destroying auth session.');

    authSessionPropagation?.cleanUp();

    if (hooks.logout) {
        await hooks.logout();
    } else {
        await storage?.clear(id);

        await store.setSession(setUnauthenticatedSession);

        await runOnAuthStateChangedCallbacks<AuthInfo>(id);
    }

    instance.onAuthStateChanged.callbacks.clear();

    await allTasksSettled(id);
    deleteInstanceTasks(id);

    const keys = Object.keys(instance) as (keyof BearAuth<AuthInfo>)[];

    for (const key of keys) {
        delete instance[key];
    }

    instances.delete(id);
}
