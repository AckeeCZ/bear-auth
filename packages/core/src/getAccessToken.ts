import { isExpired } from '~/autoRefreshToken';
import { type BearAuth } from '~/create';

import { getInstance } from './instances';
import { onAuthStateChanged } from './onAuthStateChanged';

async function onResolveAuthState(instanceId: BearAuth<unknown>['id']) {
    let unsubscribe = () => {};

    await new Promise<void>(resolve => {
        unsubscribe = onAuthStateChanged(instanceId, session => {
            if (session.status !== 'loading') {
                resolve();
            }
        });
    });

    unsubscribe();
}

/**
 * Get the access token:
 * 1. If session is 'loading', wait for it to resolve.
 * 2. If session is 'unauthenticated', return null.
 * 3. If session is 'authenticated', check if the token is expired.
 * 4. If token is expired and refresh token is available, refresh the token.
 * 5. Return the access token.
 * @param instanceId - return value of `create` method
 */
export async function getAccessToken(
    instanceId: BearAuth<unknown>['id'],
    { forceRefresh = false }: { forceRefresh?: boolean } = {},
) {
    let instance = getInstance(instanceId);

    if (instance.state.session.status === 'loading') {
        instance.logger.debug('getAccessToken', 'Waiting for loading state to resolve...');
        await onResolveAuthState(instanceId);
    }

    instance = getInstance(instanceId);

    const { status, data } = instance.state.session;

    instance.logger.debug('getAccessToken', 'Session status:', instance.state.session);

    if (status === 'unauthenticated') {
        return null;
    }

    if (status === 'authenticated') {
        if (isExpired(data.expiration) && !data.refreshToken) {
            instance.logger.debug(
                'getAccessToken',
                'Access token expired and no refresh token available. Returning null.',
            );
            return null;
        }

        if ((isExpired(data.expiration) || forceRefresh) && instance.hooks.refreshToken) {
            instance.logger.debug(
                'getAccessToken',
                'Access token expired or force refresh enabled. Refreshing access token...',
            );
            instance = await instance.hooks.refreshToken();
            instance.logger.debug('getAccessToken', 'Access token refreshed.', instance.state.session);
        }

        instance.logger.debug('getAccessToken', 'Returning access token:', instance.state.session.data!.accessToken);

        return instance.state.session.data!.accessToken;
    }

    instance.logger.debug('getAccessToken', 'Session status is not recognized. Returning null.');

    return null;
}
