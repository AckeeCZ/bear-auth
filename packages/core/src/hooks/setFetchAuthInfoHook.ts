import { stopTokenAutoRefresh } from '../autoRefreshToken.ts';
import { type BearAuth } from '../create.ts';
import { BearAuthError } from '../errors.ts';
import { getInstance } from '../instances.ts';
import { runOnAuthStateChangedCallbacks } from '../onAuthStateChanged.ts';
import { persistAuthSession } from '../storage.ts';
import { setUnauthenticatedSession, type AuthenticatedSession, type Session } from '../store/session.ts';
import { MAX_RETRY_COUNT, resolveRetry, type Retry } from './utils/retry.ts';

export type FetchAuthInfoHook<AuthInfo> = {
    handler: (authSession: AuthenticatedSession<AuthInfo>['data']) => Promise<AuthInfo>;
    action: () => Promise<Session<AuthInfo>>;
};

export type AuthData<AuthInfo> = AuthenticatedSession<AuthInfo>['data'];

/**
 * - During configuration of BearAuth, set a function to fetch the auth info from your API.
 * - Required when the `authenticaion` method returns `authInfo`.
 * @param id - return value of `create` method
 * @param handler - function to fetch the auth info
 * @returns - function to refetch the auth info
 */
export function setFetchAuthInfoHook<
    AuthInfo,
    AuthHook extends FetchAuthInfoHook<AuthInfo> = FetchAuthInfoHook<AuthInfo>,
>(
    id: BearAuth<AuthInfo>['id'],
    handler: AuthHook['handler'],
    options?: {
        retry: Retry;
    },
): AuthHook['action'] {
    async function fetchAuthInfo(retrievedAuthSession?: AuthData<AuthInfo>, failureCount = 0) {
        const instance = getInstance<AuthInfo>(id);

        const { session } = instance.state;

        if (!retrievedAuthSession && session.status !== 'authenticated') {
            throw new BearAuthError('bear-auth/not-authenticated', `Can't fetch auth data. No auth sesssion active.`);
        }

        try {
            const authSession = retrievedAuthSession ?? (session.data as AuthData<AuthInfo>);

            instance.logger.debug('[fetchAuthInfo]', 'Fetching auth data...', authSession);

            const authInfo = await handler(authSession);

            session.data = {
                ...authSession,
                authInfo,
            };

            if (!retrievedAuthSession) {
                await persistAuthSession<AuthInfo>(id);
            }

            instance.logger.debug('[fetchAuthInfo]', 'Auth data has been fetched:', authInfo);

            return getInstance<AuthInfo>(id).state.session;
        } catch (error) {
            instance.logger.error(error);

            failureCount++;

            if ((await resolveRetry(options?.retry, error, failureCount)) && failureCount < MAX_RETRY_COUNT) {
                return fetchAuthInfo(retrievedAuthSession, failureCount);
            } else {
                stopTokenAutoRefresh<AuthInfo>(id);

                setUnauthenticatedSession<AuthInfo>(id);

                await instance.storage?.clear(id);

                await runOnAuthStateChangedCallbacks<AuthInfo>(id);

                throw new BearAuthError('bear-auth/fetch-auth-data-failed', 'Failed to fetch auth data.', error);
            }
        }
    }

    const instance = getInstance<AuthInfo>(id);

    instance.hooks.fetchAuthInfo = fetchAuthInfo;

    async function refreshAuthInfo() {
        return await fetchAuthInfo();
    }

    return refreshAuthInfo;
}
