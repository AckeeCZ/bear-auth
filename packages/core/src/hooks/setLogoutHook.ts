import { stopTokenAutoRefresh } from '../autoRefreshToken.ts';
import { type BearAuth } from '../create.ts';
import { BearAuthError } from '../errors.ts';
import { getInstance } from '../instances.ts';
import { runOnAuthStateChangedCallbacks } from '../onAuthStateChanged.ts';
import {
    setSigningOutSession,
    setUnauthenticatedSession,
    type AuthenticatedSession,
    type SigningOutSession,
} from '../store/session.ts';
import { registerTask } from '../tasks.ts';
import { MAX_RETRY_COUNT, resolveRetry, type Retry } from './utils/retry.ts';

type AuthData<AuthInfo> = AuthenticatedSession<AuthInfo>['data'];

export type LogoutHook<AuthInfo> = {
    handler: (authSession: AuthData<AuthInfo>) => Promise<void>;
    action: () => Promise<void>;
};

/**
 * - Set a function to make API call to your app's backend and signs-out the user.
 * - It's always required.
 * @param id - return value of `create` method
 * @param handler - function to logout the user
 * @returns - trigger the logout process
 */
export function setLogoutHook<AuthInfo, AuthHook extends LogoutHook<AuthInfo> = LogoutHook<AuthInfo>>(
    id: BearAuth<AuthInfo>['id'],
    handler: AuthHook['handler'],
    options?: {
        retry: Retry;
    },
): AuthHook['action'] {
    async function logout(failureCount = 0) {
        const { continueWhenOnline, logger, store, storage } = getInstance<AuthInfo>(id);

        if (store.getSession().status !== 'authenticated') {
            return;
        }

        await store.setSession(setSigningOutSession);

        await runOnAuthStateChangedCallbacks<AuthInfo>(id);

        try {
            logger.debug('[logout]', 'Signing-out...');

            await stopTokenAutoRefresh<AuthInfo>(id);

            await continueWhenOnline('logout');

            await handler(store.getSession().data as SigningOutSession<AuthInfo>['data']);

            logger.debug('[logout]', 'Signed out.');
        } catch (error) {
            logger.error(error);

            failureCount++;

            if ((await resolveRetry(options?.retry, error, failureCount)) && failureCount < MAX_RETRY_COUNT) {
                return logout(failureCount);
            } else {
                throw new BearAuthError('bear-auth/logout-failed', 'Failed to logout.', error);
            }
        } finally {
            await store.setSession(setUnauthenticatedSession);

            await storage?.clear(id);

            await runOnAuthStateChangedCallbacks<AuthInfo>(id);
        }
    }

    getInstance<AuthInfo>(id).hooks.logout = logout;

    return registerTask<AuthInfo, 'logout', AuthHook['action']>(id, 'logout', logout);
}
