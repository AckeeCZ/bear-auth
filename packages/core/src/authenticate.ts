import { startTokenAutoRefresh } from './autoRefreshToken.ts';
import { type BearAuth } from './create.ts';
import { BearAuthError } from './errors.ts';
import { getExpirationTimestampWithBuffer } from './expiration.ts';
import { getInstance } from './instances.ts';
import { runOnAuthStateChangedCallbacks } from './onAuthStateChanged.ts';
import { persistAuthSession } from './storage.ts';
import { setAuthenticatedSession } from './store/session.ts';
import { registerTask } from './tasks.ts';

export type AuthenticateProps<AuthInfo> = {
    /**
     * Access token
     */
    accessToken: string;

    /**
     * Auth data (e.g. user info)
     */
    authInfo?: AuthInfo | null;

    /**
     * Expiration timestamp in ISO format.
     */
    expiration?: string | null;

    /**
     * Refresh token
     */
    refreshToken?: string | null;
};

export async function authenticateInner<AuthInfo>(
    id: BearAuth<AuthInfo>['id'],
    { accessToken, expiration = null, refreshToken = null, authInfo = null }: AuthenticateProps<AuthInfo>,
) {
    const { hooks, store } = getInstance<AuthInfo>(id);

    const refreshTokenHookRequired = Boolean(refreshToken);
    const fetchAuthInfoHookRequired = Boolean(authInfo);

    getInstance<AuthInfo>(id).flags.autoRefreshAccessTokenEnabled = Boolean(refreshTokenHookRequired && expiration);

    if (refreshTokenHookRequired && !hooks.refreshToken) {
        throw new BearAuthError(
            'bear-auth/unset-hook',
            `The 'refreshToken' property provided but the refresh token hook has not been set yet. Call setRefreshTokenHook(...) first.`,
        );
    }

    if (fetchAuthInfoHookRequired && !hooks.fetchAuthInfo) {
        throw new BearAuthError(
            'bear-auth/unset-hook',
            `The 'authInfo' property provided but the fetch authenticated data hook has not been set yet. Call setFetchAuthInfo(...) first.`,
        );
    }

    await store.setSession(() =>
        setAuthenticatedSession({
            accessToken,
            expiration: getExpirationTimestampWithBuffer(expiration),
            refreshToken,
            authInfo,
        }),
    );

    await persistAuthSession<AuthInfo>(id, store.getSession().data);

    await startTokenAutoRefresh(id);

    await runOnAuthStateChangedCallbacks<AuthInfo>(id);

    return store.getSession();
}

/**
 * Once user signs-in / signs-up, this method is called to pass the authentication data to the library.
 * - Only `accessToken` is required to be returned.
 * - `expiration` and `refreshToken` are optional, but if provided, the library will automatically refresh the access token when it expires.
 * - Don't forget to `setRefreshTokenHook` before calling this method if you want to use the refresh token.
 * - Similarly, don't forget to `setFetchAuthInfo` before calling this method if you want to use the `authInfo` property.
 * @param id - return value of `create` method
 * @param props - authentication data
 */
export function authenticate<AuthInfo>(id: BearAuth<AuthInfo>['id'], props: AuthenticateProps<AuthInfo>) {
    return registerTask<AuthInfo, 'authenticate', typeof authenticateInner<AuthInfo>>(
        id,
        'authenticate',
        authenticateInner,
    )(id, props);
}
