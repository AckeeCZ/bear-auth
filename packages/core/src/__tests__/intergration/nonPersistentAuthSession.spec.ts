import { describe, expect, test, vi } from 'vitest';

import { authenticate } from '../../authenticate.ts';
import { create } from '../../create.ts';
import { destroy } from '../../destroy.ts';
import { BearAuthError } from '../../errors.ts';
import { getExpirationTimestampWithBuffer } from '../../expiration.ts';
import { getAccessToken } from '../../getAccessToken.ts';
import { setFetchAuthInfoHook } from '../../hooks/setFetchAuthInfoHook.ts';
import { setLogoutHook } from '../../hooks/setLogoutHook.ts';
import { setRefreshTokenHook } from '../../hooks/setRefreshTokenHook.ts';
import { instances } from '../../instances.ts';
import { setContinueWhenOnline } from '../../network.ts';
import type { Session } from '../../store/session.ts';

describe('Non-Persistent Authentication Session Flows', () => {
    test('should authenticate only with an access token', async () => {
        const id = 'bear-auth-browser-test-1';
        create({ id });

        const accessToken = 'test-access-token';

        const logoutHandler = vi.fn();
        const logout = setLogoutHook(id, logoutHandler);

        const continueWhenOnline = vi.fn();
        setContinueWhenOnline(id, continueWhenOnline);

        const result = await authenticate(id, { accessToken });

        expect(result).toEqual({
            status: 'authenticated',
            data: {
                accessToken,
                authInfo: null,
                refreshToken: null,
                expiration: null,
            },
        } satisfies typeof result);

        expect(instances.get(id)?.state.session).toEqual({
            status: 'authenticated',
            data: {
                accessToken,
                authInfo: null,
                refreshToken: null,
                expiration: null,
            },
        } satisfies Session<unknown>);

        expect(await getAccessToken(id)).toBe(accessToken);

        await logout();

        expect(await getAccessToken(id)).toBeNull();

        expect(logoutHandler).toHaveBeenCalledTimes(1);
        expect(logoutHandler).toHaveBeenCalledWith({
            accessToken,
            authInfo: null,
            refreshToken: null,
            expiration: null,
        });

        expect(instances.get(id)?.state.session).toEqual({
            status: 'unauthenticated',
            data: null,
        } satisfies Session<unknown>);

        expect(continueWhenOnline).toHaveBeenCalledTimes(1);
        expect(continueWhenOnline).toHaveBeenCalledWith('logout');

        await destroy(id);
    });

    test('should authenticate with an access token and expiration', async () => {
        const id = 'bear-auth-browser-test-2';
        create({ id });

        const accessToken = 'test-access-token';
        const expiration = new Date(Date.now() + 3600 * 1000).toISOString(); // 1 hour from now

        const logoutHandler = vi.fn();
        const logout = setLogoutHook(id, logoutHandler);

        const result = await authenticate(id, {
            accessToken,
            expiration,
        });

        expect(result.status).toBe('authenticated');
        expect(result.data).toEqual({
            accessToken,
            authInfo: null,
            refreshToken: null,
            expiration: getExpirationTimestampWithBuffer(expiration),
        } satisfies typeof result.data);

        expect(await getAccessToken(id)).toBe(accessToken);

        await logout();

        expect(logoutHandler).toBeCalledTimes(1);
        expect(await getAccessToken(id)).toBeNull();

        await destroy(id);
    });

    test('should authenticate with an access token, expiration, refresh token', async () => {
        const id = 'bear-auth-browser-test-3';
        create({ id });

        const accessToken = 'test-access-token';
        const expiration = new Date(Date.now() + 3600 * 1000).toISOString(); // 1 hour from now
        const refreshToken = 'test-refresh-token';
        const freshAccessToken = 'fresh-access-token';

        const logoutHandler = vi.fn();
        const logout = setLogoutHook(id, logoutHandler);

        const continueWhenOnline = vi.fn();
        setContinueWhenOnline(id, continueWhenOnline);

        const refreshTokenHandler = vi.fn(async () => {
            return {
                accessToken: freshAccessToken,
                expiration,
                refreshToken,
            };
        });
        const refreshTokenTrigger = setRefreshTokenHook(id, refreshTokenHandler);

        expect(refreshTokenTrigger).instanceOf(Function);

        const authResult = await authenticate(id, {
            accessToken,
            expiration,
            refreshToken,
        });

        expect(authResult.status).toBe('authenticated');
        expect(authResult.data).toEqual({
            accessToken,
            authInfo: null,
            refreshToken,
            expiration: getExpirationTimestampWithBuffer(expiration),
        } satisfies typeof authResult.data);

        expect(await getAccessToken(id)).toBe(accessToken);

        // Call refresh token trigger without forcing a new access token if the token has not expired yet
        const refreshResult = await refreshTokenTrigger();

        expect(refreshTokenHandler).toBeCalledTimes(0);

        expect(refreshResult).toEqual({
            status: 'authenticated',
            data: {
                accessToken,
                expiration: getExpirationTimestampWithBuffer(expiration),
                refreshToken,
                authInfo: null,
            },
        } satisfies typeof refreshResult);

        // Call refresh token trigger with forcing a new access token
        const forcedRefreshResult = await refreshTokenTrigger({ forceRefresh: true });
        expect(refreshTokenHandler).toBeCalledTimes(1);
        expect(forcedRefreshResult).toEqual({
            status: 'authenticated',
            data: {
                accessToken: freshAccessToken,
                expiration: getExpirationTimestampWithBuffer(expiration),
                refreshToken,
                authInfo: null,
            },
        } satisfies typeof forcedRefreshResult);

        expect(continueWhenOnline).toHaveBeenCalledTimes(1);
        expect(continueWhenOnline).toHaveBeenCalledWith('refreshToken');

        expect(await getAccessToken(id)).toBe(freshAccessToken);

        await logout();

        expect(logoutHandler).toBeCalledTimes(1);
        expect(await getAccessToken(id)).toBeNull();

        expect(continueWhenOnline).toHaveBeenCalledTimes(2);
        expect(continueWhenOnline).toHaveBeenCalledWith('logout');

        await destroy(id);
    });

    test('should not authenticate with an access token and invalid expiration', async () => {
        const id = 'bear-auth-browser-test-4';
        create({ id });

        const accessToken = 'test-access-token';
        const expiration = new Date(Date.now() - 3600 * 1000).toISOString(); // 1 hour in the past

        await expect(
            authenticate(id, {
                accessToken,
                expiration,
            }),
        ).rejects.toThrowError(BearAuthError<'bear-auth/invalid-expires-in'>);

        expect(instances.get(id)?.state.session).toEqual({
            status: 'retrieving',
            data: null,
        } satisfies Session<unknown>);

        await destroy(id);
    });

    test('should authenticate with access token, expiration, refresh token, and authInfo', async () => {
        const id = 'bear-auth-browser-test-5';
        create({ id });

        const continueWhenOnline = vi.fn();
        setContinueWhenOnline(id, continueWhenOnline);

        const accessToken = 'test-access-token';
        const expiration = new Date(Date.now() + 3600 * 1000).toISOString(); // 1 hour from now
        const refreshToken = 'test-refresh-token';
        const authInfo = { userId: 'test-user-id' };

        await expect(
            authenticate(id, { accessToken, expiration, refreshToken, authInfo }),
        ).rejects.toThrowErrorMatchingSnapshot('unset-refresh-token-hook');

        const freshAccessToken = 'fresh-access-token';

        setRefreshTokenHook(id, async () => {
            return {
                accessToken: freshAccessToken,
                expiration: getExpirationTimestampWithBuffer(expiration)!,
                refreshToken,
                authInfo,
            };
        });

        await expect(
            authenticate(id, { accessToken, expiration, refreshToken, authInfo }),
        ).rejects.toThrowErrorMatchingSnapshot('unset-fetch-auth-info-hook');

        const freshAuthInfo = { userId: 'fresh-user-id' };
        const authInfoHandler = vi.fn(async () => {
            return freshAuthInfo;
        });
        const fetchAuthInfo = setFetchAuthInfoHook(id, authInfoHandler);

        const authResult = await authenticate(id, { accessToken, expiration, refreshToken, authInfo });

        expect(authResult.status).toBe('authenticated');
        expect(authResult.data).toEqual({
            accessToken,
            expiration: getExpirationTimestampWithBuffer(expiration),
            refreshToken,
            authInfo,
        } satisfies typeof authResult.data);

        expect(authInfoHandler).toBeCalledTimes(0);

        expect(await fetchAuthInfo()).toEqual({
            status: 'authenticated',
            data: {
                accessToken,
                expiration: getExpirationTimestampWithBuffer(expiration),
                refreshToken,
                authInfo: freshAuthInfo,
            },
        } satisfies Session<unknown>);

        expect(authInfoHandler).toBeCalledTimes(1);

        expect(continueWhenOnline).toHaveBeenCalledTimes(1);
        expect(continueWhenOnline).toHaveBeenCalledWith('fetchAuthInfo');

        expect(await getAccessToken(id, { forceRefresh: true })).toBe(freshAccessToken);

        await destroy(id);
    });
});
