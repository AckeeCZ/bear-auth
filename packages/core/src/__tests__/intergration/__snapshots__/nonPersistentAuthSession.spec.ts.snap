// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Non-Persistent Authentication Session Flows > should authenticate with access token, expiration, refresh token, and authInfo > unset-fetch-auth-info-hook 1`] = `[BearAuthError: The 'authInfo' property provided but the fetch authenticated data hook has not been set yet. Call setFetchAuthInfo(...) first.]`;

exports[`Non-Persistent Authentication Session Flows > should authenticate with access token, expiration, refresh token, and authInfo > unset-refresh-token-hook 1`] = `[BearAuthError: The 'refreshToken' property provided but the refresh token hook has not been set yet. Call setRefreshTokenHook(...) first.]`;
